File: chat/templates.go
package chat

import (
	"encoding/json"
	"os"
	"path/filepath"
)

// Template is a pre-made chat question with a display name and the actual prompt sent to the LLM.
type Template struct {
	Name   string `json:"name"`
	Prompt string `json:"prompt"`
}

var defaultTemplates = []Template{
	{
		Name:   "TL;DR",
		Prompt: "Give me a brief summary (tl;dr) of this recording in 2-3 sentences.",
	},
	{
		Name:   "What did I miss?",
		Prompt: "What were the most important points discussed? List the key decisions, action items, or insights I should know about.",
	},
	{
		Name:   "Action items",
		Prompt: "List all action items, tasks, or commitments mentioned in this recording. Include who is responsible if mentioned.",
	},
	{
		Name:   "Key decisions",
		Prompt: "What decisions were made during this meeting or conversation? List them clearly.",
	},
	{
		Name:   "Open questions",
		Prompt: "What questions were raised but not resolved? List any unresolved topics or follow-up items.",
	},
}

// DefaultTemplates returns the built-in set of question templates.
func DefaultTemplates() []Template {
	return defaultTemplates
}

// LoadTemplates reads templates from a JSON file.
func LoadTemplates(path string) ([]Template, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var templates []Template
	if err := json.Unmarshal(data, &templates); err != nil {
		return nil, err
	}
	return templates, nil
}

// EnsureTemplatesFile loads templates from path, creating the file with defaults if it doesn't exist.
func EnsureTemplatesFile(path string) []Template {
	templates, err := LoadTemplates(path)
	if err == nil {
		return templates
	}
	defaults := DefaultTemplates()
	data, err := json.MarshalIndent(defaults, "", "  ")
	if err != nil {
		return defaults
	}
	if mkErr := os.MkdirAll(filepath.Dir(path), 0755); mkErr == nil {
		_ = os.WriteFile(path, data, 0644)
	}
	return defaults
}

File: src/llm/templates.rs
use std::path::Path;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Template {
    pub name: String,
    pub prompt: String,
}

fn default_templates() -> Vec<Template> {
    let t = |name: &str, prompt: &str| Template {
        name: name.to_string(),
        prompt: prompt.to_string(),
    };
    vec![
        t(
            "TL;DR",
            "Give me a brief summary (tl;dr) of this recording in 2-3 sentences.",
        ),
        t(
            "What did I miss?",
            "What were the most important points discussed? List the key decisions, action items, or insights I should know about.",
        ),
        t(
            "Action items",
            "List all action items, tasks, or commitments mentioned in this recording. Include who is responsible if mentioned.",
        ),
        t(
            "Key decisions",
            "What decisions were made during this meeting or conversation? List them clearly.",
        ),
        t(
            "Open questions",
            "What questions were raised but not resolved? List any unresolved topics or follow-up items.",
        ),
    ]
}

pub fn load_templates(path: &Path) -> anyhow::Result<Vec<Template>> {
    let data = std::fs::read_to_string(path)?;
    Ok(serde_json::from_str(&data)?)
}

/// Load templates, creating the file with defaults on any failure (Go parity:
/// malformed JSON also resets to defaults).
pub fn ensure_templates_file(path: &Path) -> Vec<Template> {
    if let Ok(templates) = load_templates(path) {
        return templates;
    }
    let defaults = default_templates();
    if let Some(dir) = path.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    if let Ok(json) = serde_json::to_string_pretty(&defaults) {
        let _ = std::fs::write(path, json);
    }
    defaults
}
